staging/src/k8s.io/pod-security-admission/admission/admission.go

```diff
@@ -0,0 +1,537 @@
+			api.CompareLevels(policy.Enforce.Level, oldPolicy.Enforce.Level) < 1 {
+			return allowedResponse()
+		}
+	} else if err != nil { // Create
+		return invalidResponse(err.Error())
+	}
+
+	// We only need to test pods if the enforce policy is not privileged.
+	if policy.Enforce.Level == api.LevelPrivileged {
+		return allowedResponse()
+	}
+
+	response := allowedResponse()
+	response.Warnings = a.CheckPodsInNamespace(ctx, namespace.Name, policy.Enforce)
+
+	return response
+}
+
+func (a *Admission) ValidatePod(ctx context.Context, attrs admission.Attributes) admissionv1.AdmissionResponse {
+	// short-circuit on subresources other than ephemeralcontainers
+	if attrs.GetSubresource() != "" && attrs.GetSubresource() != "ephemeralcontainers" {
+		return allowedResponse()
+	}
+	// short-circuit on exempt namespaces and users
+	if a.exemptNamespace(attrs.GetNamespace()) || a.exemptUser(attrs.GetUserInfo().GetName()) {
+		return allowedResponse()
+	}
+
+	pod, ok := attrs.GetObject().(*corev1.Pod)
+	if !ok {
+		klog.InfoS("failed to assert pod type", "type", reflect.TypeOf(attrs.GetObject()))
+		return internalErrorResponse("failed to decode pod")
+	}
+	enforce := true
+	if attrs.GetOperation() == admission.Update {
+		oldPod, ok := attrs.GetOldObject().(*corev1.Pod)
+		if !ok {
+			klog.InfoS("failed to assert old pod type", "type", reflect.TypeOf(attrs.GetOldObject()))
+			return internalErrorResponse("failed to decode old pod")
+		}
+		if !isSignificantPodUpdate(pod, oldPod) {
+			// Nothing we care about changed, so always allow the update.
+			enforce = false
+		}
+	}
+	return a.CheckPod(ctx, &pod.ObjectMeta, &pod.Spec, enforce)
+}
+
+func (a *Admission) ValidatePodController(ctx context.Context, attrs admission.Attributes) admissionv1.AdmissionResponse {
+	// short-circuit on subresources
+	if attrs.GetSubresource() != "" {
+		return allowedResponse()
+	}
+	// short-circuit on exempt namespaces and users
+	if a.exemptNamespace(attrs.GetNamespace()) || a.exemptUser(attrs.GetUserInfo().GetName()) {
+		return allowedResponse()
+	}
+
+	podMetadata, podSpec, err := a.PodSpecExtractor.ExtractPodSpec(attrs.GetObject())
+	if err != nil {
+		klog.ErrorS(err, "failed to extract pod spec")
+		return internalErrorResponse("failed to extract pod template")
+	}
+	if podMetadata == nil && podSpec == nil {
+		// if a controller with an optional pod spec does not contain a pod spec, skip validation
+		return allowedResponse()
+	}
+	return a.CheckPod(ctx, podMetadata, podSpec, false)
+}
+
+// CheckPod looks up the policy for the pods namespace, and checks it against the given pod(-like) object.
+// The enforce policy is only checked if enforce=true.
+func (a *Admission) CheckPod(ctx context.Context, podMetadata *metav1.ObjectMeta, podSpec *corev1.PodSpec, enforce bool) admissionv1.AdmissionResponse {
+	// short-circuit on exempt runtimeclass
+	if a.exemptRuntimeClass(podSpec.RuntimeClassName) {
```

Comment:
I haven't read yet, but I'm betting this is here because namespace admission reuses this method.  There's a unit test ensuring someone doesn't optimize this into the pod evaluation method?

Reply:
Both pod-controller and pod validation route here. Namespace evaluation was missing this. Added and added a unit test.